nestjs/src/courses/task-verifications/task-verifications.service.ts

```diff
@@ -67,4 +87,82 @@ export class TaskVerificationsService {
       throw new BadRequestException('The answers cannot be checked if there were no attempts.');
     }
   }
+
+  public async createTaskVerification(
+    courseTaskId: number,
+    studentId: number,
+    data: { githubId: string; body: any },
+  ): Promise<{ id?: number }> {
+    const [courseTask, student] = await Promise.all([
+      this.courseTasksRepository.findOneByOrFail({ id: courseTaskId }),
+      this.studentsRepository.findOneByOrFail({ id: studentId }),
+    ]);
+
+    if (courseTask == null || student == null) {
+      throw new BadRequestException('No student or not valid course task');
+    }
+
+    if (courseTask.courseId !== student.courseId) {
+      throw new BadRequestException(`Course task does not belong to the student's course`);
+    }
+
+    const existing = await this.taskVerificationsRepository.findOne({
+      where: {
+        status: 'pending',
+        studentId,
+        courseTaskId,
+        updatedDate: MoreThan(dayjs().add(-1, 'hour').toDate()),
+      },
+      select: ['id'],
+    });
+
+    if (existing != null) {
+      throw new HttpException(`Task verification [${existing.id}] is in progress`, HttpStatus.TOO_MANY_REQUESTS);
+    }
+
+    const expired = await this.courseTasksRepository.findOne({
+      where: {
+        id: courseTaskId,
+        studentEndDate: MoreThan(dayjs().toDate()),
+      },
+      select: ['id'],
+    });
+
+    if (expired != null) {
+      throw new HttpException(`Task Verification [${expired.id}] expired`, 423);
+    }
+
+    if (courseTask.type === TaskType.SelfEducation) {
+      await this.seflEducationService.createSelfeducationVerification({
+        courseId: courseTask.courseId,
+        courseTask,
+        studentId: student.id,
+        studentAnswers: data.body,
+      });
+      return { id: undefined };
```

Comment:
why `null` is better than `undefined`?

Reply:
IMO `null` is better if we intentionally set empty value there while `undefined` refers to the situation when something went wrong. But my point was to return `null` instead of `{ id: undefined }`, not `{ id: null }`

Reply:
I'm not sure why we have variable situation (as I've not developed this method) but returning different payload from REST API will not be good for client when you can get object or nothing. REST API response format should be consistent

Reply:
Maybe we should refactor this later, ok